src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/EnumConverter.cs

```diff
@@ -60,7 +60,16 @@ public EnumConverter(EnumConverterOptions converterOptions, JsonNamingPolicy? na
 
                 T value = (T)values.GetValue(i)!;
                 ulong key = ConvertToUInt64(value);
-                string name = names[i];
+                string name;
+
+                if (serializerOptions.DictionaryKeyPolicy != null)
+                {
+                    name = serializerOptions.DictionaryKeyPolicy.ConvertName(names[i]);
+                }
+                else
+                {
+                    name = names[i];
+                }
```

Comment:
Hmm this transformation shouldn't be applied to the main cache which honors the `namingPolicy` passed into the constructor. The transformation should be done lazily in the `WriteWithQuotes` method below. That method handles the serialization of enums as dictionary keys. For better throughput, you should cache these transformations in a new, separate cache.

In light of this, please add some tests to ensure that `JsonSerializerOptions.DictionaryKeyPolicy` is not applied when enums are serialized..

- as root level types e.g. `JsonSerializer.Serialize(DayOfWeek.Friday, options);`
- as POCO properties e.g. `JsonSerializer.Serialize(new ClassWithEnumProperty(), options);`
- as collection elements e.g. `JsonSerializer.Serialize(new List<DayOfWeek> { ... }, options);` or `JsonSerializer.Serialize(new Dictionary<DayOfWeek, TValue> { ... }, options);`

Reply:
Thank you for the comment, I had been thinking about the implementation (except the second cache thing), but ended up like this. I'll fix it (as you suggested) as soon as possible.